service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/CatalogHandlerUtils.java

```diff
@@ -433,39 +446,174 @@ protected TableMetadata commit(TableOperations ops, UpdateTableRequest request)
+                    throw new ValidationFailureException(e);
+                  }
+
+                  // snapshot-id the client expects the table current_snapshot_id
+                  long expectedCurrentSnapshotId = assertRefSnapshotId.snapshotId();
+                  // table current_snapshot_id.
+                  long currentSnapshotId = base.currentSnapshot().snapshotId();
+                  List<MetadataUpdate> metadataUpdates =
+                      generateUpdatesToRemoveNoopSnapshot(
+                          base,
+                          currentSnapshotId,
+                          expectedCurrentSnapshotId,
+                          setSnapshotRef.name());
+
+                  if (metadataUpdates == null || metadataUpdates.isEmpty()) {
+                    // Nothing can be done as this implies that there were not all
+                    // No-op snapshots (REPLACE) between expectedCurrentSnapshotId and
+                    // currentSnapshotId. hence re-throw the exception caught.
+                    throw new ValidationFailureException(e);
+                  }
+
+                  // Set back the ref we wanted to set, back to the snapshot-id
+                  // the client is expecting the table to be at.
+                  metadataBuilder.setBranchSnapshot(
+                      expectedCurrentSnapshotId, setSnapshotRef.name());
+
+                  // apply the remove snapshots update in the current metadata.
+                  // NOTE: we need to setRef to expectedCurrentSnapshotId first and then apply
+                  // remove, as otherwise the remove will drop the reference.
+                  // NOTE: we can skip removing the now orphan base. Its not a hard requirement.
+                  // just something good to do, and not leave for Remove Orphans.
+                  metadataUpdates.forEach((update -> update.applyTo(metadataBuilder)));
+                  // Ref rolled back update correctly to snapshot to be committed parent now.
+                  newBase = metadataBuilder.build();
+                  // move the lastSequenceNumber back, to apply snapshot properly on the
+                  // current-metadata Seq number are considered increasing monotonically
+                  // snapshot over snapshot, the client generates the manifest list and hence
+                  // the sequence number can't be changed for a snapshot the only possible option
+                  // then is to change the sequenceNumber tracked by metadata.json
+                  Class<?> clazz = newBase.getClass();
+                  try {
+                    Field field = clazz.getDeclaredField("lastSequenceNumber");
+                    field.setAccessible(true);
+                    // this should point to the sequence number that current tip of the
+                    // branch belongs to, as the new commit will be applied on top of this.
+                    field.set(newBase, newBase.currentSnapshot().sequenceNumber());
```

Comment:
Is it critical to have contiguous sequence numbers? When we save the new data into storage, some sequence number will be reused without proper branches. :thinking:

Reply:
Why not use the `Builder`, transfer snapshots manually, and let the builder figure out the sequence number.